platform/distance.cpp

```diff
@@ -0,0 +1,217 @@
+  case Distance::Units::Kilometers:
+    return {measurement_utils::MilesToMeters(distance) / 1000, Distance::Units::Kilometers};
+  case Distance::Units::Feet:
+    return {measurement_utils::MilesToFeet(distance), Distance::Units::Feet};
+  default: UNREACHABLE();
+  }
+}
+
+double WithPrecision(double value, uint8_t precision)
+{
+  double const factor = std::pow(10, precision);
+  return std::round(value * factor) / factor;
+}
+}  // namespace
+
+Distance::Distance() : Distance(0.0) {}
+
+Distance::Distance(double distanceInMeters) : Distance(distanceInMeters, Units::Meters) {}
+
+Distance::Distance(double distance, platform::Distance::Units units)
+  : m_distance(distance), m_units(units)
+{
+}
+
+Distance Distance::CreateFormatted(double distanceInMeters)
+{
+  return Distance(distanceInMeters).ToPlatformUnitsFormatted();
+}
+
+Distance Distance::CreateAltitudeFormatted(double meters)
+{
+  Distance elevation = Distance(meters).To(
+      measurement_utils::GetMeasurementUnits() == measurement_utils::Units::Metric ? Units::Meters
+                                                                                   : Units::Feet);
+  if (elevation.IsLowUnits())
+    elevation.m_distance = WithPrecision(elevation.m_distance, 0);
+  return elevation;
+}
+
+bool Distance::IsValid() const { return m_distance >= 0.0; }
+
+bool Distance::IsLowUnits() const { return m_units == Units::Meters || m_units == Units::Feet; }
+
+bool Distance::IsHighUnits() const { return !IsLowUnits(); }
+
+Distance Distance::To(Distance::Units units) const
+{
+  if (m_units == units)
+    return *this;
+
+  switch (m_units)
+  {
+  case Units::Meters: return MetersTo(m_distance, units);
+  case Units::Kilometers: return KilometersTo(m_distance, units);
+  case Units::Feet: return FeetTo(m_distance, units);
+  case Units::Miles: return MilesTo(m_distance, units);
+  default: UNREACHABLE();
+  }
+}
+
+Distance Distance::ToPlatformUnitsFormatted() const
+{
+  return To(measurement_utils::GetMeasurementUnits() == measurement_utils::Units::Metric
+                ? Units::Meters
+                : Units::Feet)
+      .GetFormattedDistance();
+}
+
+double Distance::GetDistance() const { return m_distance; }
+
+Distance::Units Distance::GetUnits() const { return m_units; }
+
+std::string Distance::GetDistanceString() const
+{
+  std::ostringstream os;
+  os << std::defaultfloat << m_distance;
+  return os.str();
+}
+
+#if defined(OMIM_OS_IPHONE) || defined(OMIM_OS_ANDROID)
+std::string Distance::GetUnitsString() const
+{
+  switch (m_units)
+  {
+  case Units::Meters: return GetLocalizedString("meter");
+  case Units::Kilometers: return GetLocalizedString("kilometer");
+  case Units::Feet: return GetLocalizedString("foot");
+  case Units::Miles: return GetLocalizedString("mile");
+  default: UNREACHABLE();
+  }
+}
+#else
+std::string Distance::GetUnitsString() const
+{
+  switch (m_units)
+  {
+  case Units::Meters: return "m";
```

Comment:
Can longer strings above ("meter") be renamed in strings to "m" so we have the same code for all platforms?